NitroxLauncher/Models/Utils/WindowsHelper.cs

```diff
@@ -55,6 +64,119 @@ internal static void RestartAsAdmin()
+                }
+            }
+        }
+
+        internal static void CheckFirewallRules(FirewallDirection direction)
+        {
+            CheckClientFirewallRules(direction);
+            CheckServerFirewallRules(direction);
+            CheckHamachiFirewallRules(direction);
+        }
+
+        private static void CheckServerFirewallRules(FirewallDirection direction)
+        {
+            string serverPath = Path.Combine(Path.GetDirectoryName(Assembly.GetEntryAssembly().Location), ServerLogic.SERVER_EXECUTABLE);
+
+            AddExclusiveFirewallRule(Path.GetFileName(serverPath), serverPath, direction);
+        }
+
+        private static void CheckClientFirewallRules(FirewallDirection direction)
+        {
+            string clientPath = Path.Combine(LauncherLogic.Config.SubnauticaPath, "Subnautica.exe");
+
+            AddExclusiveFirewallRule(Path.GetFileName(clientPath), clientPath, direction);
+        }
+
+        private static void CheckHamachiFirewallRules(FirewallDirection direction)
+        {
+            static IPAddress GetHamachiAddress()
+            {
+                foreach (NetworkInterface networkInterface in NetworkInterface.GetAllNetworkInterfaces())
+                {
+                    if (networkInterface.Name == "Hamachi")
+                    {
+                        foreach (UnicastIPAddressInformation addressInfo in networkInterface.GetIPProperties().UnicastAddresses)
+                        {
+                            if (addressInfo.Address.AddressFamily == AddressFamily.InterNetwork)
+                            {
+                                return addressInfo.Address;
+                            }
+                        }
+                    }
+                }
+
+                return null;
+            }
+
+            const string NAME = "HamachiNitrox";
+
+            if (GetHamachiAddress() == null || !FirewallWASRule.IsLocallySupported)
+            {
+                return;
+            }
+
+            foreach (IFirewallRule firewallRule in FirewallManager.Instance.Rules)
+            {
+                if (firewallRule.Name == NAME)
+                {
+                    return;
+                }
+            }
+
+            FirewallWASRule rule = new(NAME, FirewallAction.Allow, direction, FirewallManager.Instance.GetActiveProfile().Type)
+            {
+                Description = "Ignore Hamachi network",
+                Protocol = FirewallProtocol.Any,
+                RemoteAddresses = new[] { new NetworkAddress(GetHamachiAddress()) },
+                IsEnable = true
+            };
+
+            FirewallManager.Instance.Rules.Add(rule);
+        }
+
+        private static void AddExclusiveFirewallRule(string name, string filePath, FirewallDirection direction)
+        {
+            if (!File.Exists(filePath))
+            {
+                throw new FileNotFoundException("Unable to add firewall rule to non-existent program", filePath);
```

Comment:
Do we want to show a popup of this error, since this error will end up being catch up by the error handler of WPF resulting in a MessageBox ?

Maybe we should just log this error and/or putting a little toaster to warn client ?

Reply:
This error can happen when the subnautica path isn't defined. Since this firewall change is secondary issue, since you won't be able to use Nitrox, I don't see it necessary and could confuse the user unless we really explain it well.